### PR TITLE
Allow trait bounds to be overridden in macro

### LINKED ISSUE #696: Allow trait bounds to be overridden in macro
First off, apologies if this is wrong; I haven't spent long digging into the macro code here!

So, currently we can generate a client impl using the proc macro like so:

```
#[rpc(client, namespace = "foo")]
pub trait MyRpc<T> {
    #[method(name = "bar")]
    fn method(&self) -> RpcResult<T>;
}
```

I believe this generates something like an `trait MyRpcClient<T>: ClientT where T: Send + Sync + 'static` trait, and an impl for it that means the RPC `Client` impls these methods (very cool btw!).

However, in some cases this general rule of applying `Send + Sync + 'static` to all type params is unnecessarily restrictive. Take the following example:

```
trait Config {
    type Hash: Send + Sync + 'static;
}

#[rpc(client, namespace = "foo")]
pub trait MyRpc<T: Config> {
    #[method(name = "bar")]
    fn method(&self) -> RpcResult<T::Hash>;
}
```

This generates something like `trait MyRpcClient<T>: ClientT where T: Send + Sync + 'static + Config`, but here `T` does not need to be `Send + Sync + 'static` since it's not actually returned or passed in anywhere (we only care about `T::Hash`).

This is quite a common issue in derives too, for instance if we take this struct:

```
#[derive(Clone)]
struct Foo<T> {
    marker: std::marker::PhantomData<T>
}
```

The generated `Clone` impl is something like `impl <T> Clone for Foo<T> where T: Clone {..}`, which is unnecessary since T really doesn't need to be cloneable (it's not instantiated anywhere).

A common solution is to allow people an escape hatch to override the automatically applied trait bounds. The `derivative` library provides this for common traits, so we could impl Clone for all `Foo<T>`'s like so:

```
#[derive(Derivative)]
#[derivative(Clone(bound = ""))]
struct Foo<T> {
    marker: std::marker::PhantomData<T>
}
```

This generates an impl more like `impl <T> Clone for Foo<T> {..}` (we override the Clone bounds to be nothing).

Serde has a similar feature:

```
#[derive(Serialize)]
#[derde(bound = "")]
struct Foo<T> {
    marker: std::marker::PhantomData<T>
}
```

Again, we can manually set the bound to something like "T: MyTrait", or in our case, nothing ("").

Perhaps it would be useful to provide a similar "escape hatch" in jsonrpsee? With such a feature, the earlier example could be made to work with something like:

```
trait Config {
    type Hash: Send + Sync + 'static;
}

#[rpc(client, namespace = "foo", client_bound = "")]
pub trait MyRpc<T: Config> {
    #[method(name = "bar")]
    fn method(&self) -> RpcResult<T::Hash>;
}
```

It would need a little thought;
- does `client_bounds` override _all_ bounds that would be applied, so that they all need to be manual? In this case we might need eg `client_bound = "T::Hash: Serialize + Deserialize"` or something?
- would we need `client_bound` and `server_bound` to allow overriding each?

### PR DESCRIPTION
**Pull request by lexnv**
This PR adds the `server_bounds` and `client_bounds` attributes of the `rpc` macro.

The attributes allow users to specify custom bounds for their traits, as a way to opt-out
of the default `Send + Sync + 'static`  where appropriate.

### Example

Without the custom provided bounds, the RPC macro would add by default 
`Conf : Send + Sync + 'static + jsonrpsee::core::DeserializeOwned` for client implementation and
`Conf : Send + Sync + 'static + jsonrpsee::core::Serialize` for server implementation.
This rule is quite restrictive, as we need just `Conf::Hash` to comply to those rules.
 
```rust
pub trait Config {
	type Hash: Send + Sync + 'static;
}

impl Config for ExampleHash {
	type Hash = Self;
}

#[rpc(server, client, namespace = "foo", client_bounds(Conf::Hash: jsonrpsee::core::DeserializeOwned), server_bounds(Conf::Hash: jsonrpsee::core::Serialize))]
pub trait Rpc<Conf: Config> {
	#[method(name = "bar")]
	fn method(&self) -> Result<Conf::Hash, Error>;
}

```

While at it, add a new example illustrating the custom bounds and a few UI tests to capture
proper error messages.

Closes #696.

### CODE REVIEW

File: examples/examples/proc_macro_bounds.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+use jsonrpsee::core::{async_trait, Error};
+use jsonrpsee::proc_macros::rpc;
+use jsonrpsee::ws_client::WsClientBuilder;
+use jsonrpsee::ws_server::{WsServerBuilder, WsServerHandle};
+
+type ExampleHash = [u8; 32];
+
+pub trait Config {
+	type Hash: Send + Sync + 'static;
```
Comment (niklasad1) [inline examples/examples/proc_macro_bounds.rs:37]: IIRC: before/"currently" we add the trait bound the entire trait right?

Can you just verify that we don't add trait bounds on associated types that are used in the RPC trait?

such as:

```rust
pub trait Config {
   type Hash: Send + Sync + 'static;
   type NotUsed;
}

```

**Review by niklasad1 — COMMENTED** [review]
great.

some UI tests that it actually work as well then this should be ready to go :)

File: examples/examples/proc_macro_bounds.rs (hunk 1 above)
Comment (lexnv) [reply examples/examples/proc_macro_bounds.rs:37]: Debugging the proc-macro, it seems that "currently" (without custom `bounds` attribute):

```rust
pub trait Config {
	type Hash: Send + Sync + 'static;
	type NotUsed;
}

#[rpc(server, namespace = "foo")]
pub trait Rpc<Conf: Config> {
	#[method(name = "bar")]
	fn method(&self) -> Result<Conf::Hash, Error>;
}
```

The `rpc` macro implies just: `Conf : Send + Sync + 'static + jsonrpsee :: core :: Serialize`.

Although, this fails to compile due to:

```rust
55  | #[rpc(server, client, namespace = "foo", client_bounds(Conf::Hash: jsonrpsee::core::DeserializeOwned))]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `<Conf as Config>::Hash`
```

Comment (niklasad1) [reply examples/examples/proc_macro_bounds.rs:37]: Yeah, that's fine I meant with the overrides for the `Hash`

```rust

pub trait Config {
	type Hash: Send + Sync + 'static;
	type NotUsed;
}

#[rpc(server, client, client_bounds(Conf::Hash: jsonrpsee::core::Serialize))]
pub trait Rpc<Conf: Config> {
	#[method(name = "bar")]
	fn method(&self) -> Result<Conf::Hash, Error>;
}
```

Such that it doesn't expand to something like:

```rust
pub trait Rpc<T: Client, C: Config> 
where 
        C::Hash: Serialize,
        C::NotUsed: Send + Sync + Serialize + DeserializeOwned
{
	#[method(name = "bar")]
	fn method(&self) -> Result<Conf::Hash, Error>;
}
```

But as it works in subxt all fine, IIRC we only inspect the parameters and return types anyway add bounds on them so all good.

**Review by niklasad1 — APPROVED** [review]
excellent, nice work

some minor suggestions on testing but really good to have this finally

**Comment by jsdw** [comment]
Just wondering; is it possible to define something like:

```rust
#[rpc(server, client, client_bounds(Conf::Hash: jsonrpsee::core::Serialize))]
pub trait Rpc<Conf: Config> {
	fn generic_input<T>(&self, input: T) -> Result<Conf::Hash, Error>;
        fn generic_output<O>(&self, input: T) -> Result<O, Error>;
}
```

And if so, can we control the bounds of `T` and `O`? This may be a dumb question :)

**Review by jsdw — APPROVED** [review]
Not sure whether the question above makes sense or not (I can't recall whether it's possible to add generic bounds in that way), but regardless, this looks great; nice job!

**Comment by lexnv** [comment]
Managed to define something similar:

```rust
#[rpc(client, client_bounds(
	Conf::Hash: jsonrpsee::core::DeserializeOwned,
	T: jsonrpsee::core::Serialize + Send + Sync + 'static,
	O: jsonrpsee::core::DeserializeOwned)
)]
pub trait Rpc2<Conf: Config, T, O> {
	#[method(name = "bar")]
	fn generic_input<T>(&self, input: T) -> Result<Conf::Hash, Error>;
	#[method(name = "foo")]
	fn generic_output<T, O>(&self, input: T) -> Result<O, Error>;
}
```
with `T: jsonrpsee::core::Serialize + Send + Sync + 'static` (as it is input for client) and `O: jsonrpsee::core::DeserializeOwned` (output)

**Comment by niklasad1** [comment]
> Just wondering; is it possible to define something like:

```rust
#[rpc(server, client, client_bounds(Conf::Hash: jsonrpsee::core::Serialize))]
pub trait Rpc<Conf: Config> {
	fn generic_input<T>(&self, input: T) -> Result<Conf::Hash, Error>;
        fn generic_output<O>(&self, input: T) -> Result<O, Error>;
}
```

> And if so, can we control the bounds of `T` and `O`? This may be a dumb question :)

The bounds on individual methods doesn't work because the server has  `fn into_rpc(self) -> jsonrpsee::RpcModule<Self>` and then these  "type param" are not `bound` to anything so it won't compile.

It's would need us to find all `individual type params` and generate `fn into_rpc<T1, ... Tn>(self) -> jsonrpsee::RpcModule<Self>
then it would work I guess.

I think the same applies to client trait too i.e, that we don't really parse `type params` on individual methods.

**Pull request merged**